lib/src/components/NxScrollRender/NxScrollRender.tsx

```diff
@@ -0,0 +1,133 @@
+      [childHeight, setChildHeight] = useState<number | null>(null),
+      normalizedChildHeight = childHeight ?? 0,
+      renderedChildCount = parentHeight == null || childHeight == null ?
+        null : Math.min(Math.ceil(divOrZero(parentHeight, childHeight)) + 2, childCount),
+
+      cloneWithKey = (child: ReactElement, idx: number) =>
+        React.cloneElement(child, { key: idx % (renderedChildCount as number) }),
+      keyedChildren = useMemo(
+          () => reuseChildren === false || isNil(renderedChildCount) ? childArray : childArray.map(cloneWithKey),
+          [childArray, reuseChildren, renderedChildCount]
+      ),
+
+      sumChildHeight = normalizedChildHeight * childCount,
+      renderedChildHeight = renderedChildCount == null ? 0 : normalizedChildHeight * renderedChildCount,
+
+      [firstRenderedChildIdx, setFirstRenderedChildIdx] = useState(0),
+
+      renderedChildCountWithDefault = renderedChildCount ?? initialChildCount ?? DEFAULT_INITIAL_CHILD_COUNT,
+      renderedRealChildren =
+          keyedChildren.slice(firstRenderedChildIdx, firstRenderedChildIdx + renderedChildCountWithDefault),
+
+      sumSpacerHeight = sumChildHeight - renderedChildHeight,
+      clampSpacerHeight = clamp(0, sumSpacerHeight),
+      leadingSpacerHeight = clampSpacerHeight(firstRenderedChildIdx * normalizedChildHeight),
+      trailingSpacerHeight = renderedChildCount == null ? 0 : clampSpacerHeight(
+          sumChildHeight - (leadingSpacerHeight + renderedChildCount * normalizedChildHeight)
+      );
+
+  const renderedChildren = childCount ? (
+    <>
+      {React.createElement(spacerEl ?? 'div', {
+        ref: leadingSpacerRefFn,
+        role: 'presentation',
+        className: 'nx-scroll-render__spacer',
+        style: { height: leadingSpacerHeight }
+      })}
+      {renderedRealChildren}
+      {React.createElement(spacerEl ?? 'div', {
+        ref: trailingSpacerRefFn,
+        role: 'presentation',
+        className: 'nx-scroll-render__spacer',
+        style: { height: trailingSpacerHeight }
+      })}
+    </>
+  ) : null;
+
+  const updateRendering = useThrottleCallback(useCallback(function updateRendering() {
+    const scrollTop = parentRef.current?.scrollTop ?? 0,
+        parentTop = parentRef.current?.getBoundingClientRect().top,
+        leadingSpacerTop = leadingSpacerRef.current?.getBoundingClientRect().top,
+        topDifference = (parentTop ?? 0) - (leadingSpacerTop ?? 0),
+        topTooClose = topDifference < normalizedChildHeight,
+        topTooFar = topDifference > normalizedChildHeight * 2,
+        adjust =
+          topTooClose ? dec :
+          topTooFar ? inc :
+          identity,
+        clampFirstRenderedChildIdx =
+            renderedChildCount == null ? always(0) : clamp(0, childCount - renderedChildCount),
+        newFirstRenderedChildIdx = clampFirstRenderedChildIdx(
+            adjust(Math.floor(divOrZero(scrollTop, normalizedChildHeight)) - 2)
```

Comment:
how does the -2 factor in here?

Reply:
> how does the -2 factor in here?

Let's just do an example.  Let's say you have children that are 10px high, and you've just (instantaneously, for the sake of simplicity) scrolled from the top of the list down to where the top of the fifth child is exactly aligned with the top of the container.  Also let's say the container is 100px from the top of the viewport, is 70px tall, and there's a total of 20 children.  So here's how I believe our variables would work out:

`scrollTop`: 50
`parentTop`: 100
`leadingSpacerTop`: 50
`topDifference`: 50 (hmm... is this always going to be the same as `scrollTop`? I will double check and optimize if possible)
`topTooClose`: false (50 is not less than 10, so we do not need to add more children at the top)
`topTooFar`: true (50 is greater than 10 times 2, so we current have too many children at the top (and accordingly, not enough at the bottom))
`adjust`: increment function
`clampFirstRenderedChildIdx`: function that clamps between 0 and 13.  9 Because the total child count is 20, and the `renderedChildCount` is the number of visible children plus 2 (notionally, one on each end) which is 7.  See line 39 where `renderedChildCount` is computed.
`newFirstRenderedChildIdx`: first we divide 50 by 10 to get 5.  Then subtract 2 (which is what you're asking about). Then call the `adjust` function which increments to 3. Then that passes the `clamp` and 3 is the final result.  

Remember though, this is an array index, so this 3 refers to the _fourth_ child.  And remember in this example that the fifth child is the first one visible at the new scroll position.  So this result in one child outside of view above the visible area, then the five visible children, and then one child outside of view below the visible area.  The children outside of view on either end are so that as you scroll there's something there to see rather than just immediately exposing the spacer.

TL;DR: I think we subtract 1 because array indexes are zero-based, and 1 more because we want the first rendered child to be the one above the first one you can actually see.

Reply:
> > how does the -2 factor in here?
> 
> Let's just do an example. Let's say you have children that are 10px high, and you've just (instantaneously, for the sake of simplicity) scrolled from the top of the list down to where the top of the fifth child is exactly aligned with the top of the container. Also let's say the container is 100px from the top of the viewport, is 70px tall, and there's a total of 20 children. So here's how I believe our variables would work out:
> 
> `scrollTop`: 50 `parentTop`: 100 `leadingSpacerTop`: 50 `topDifference`: 50 (hmm... is this always going to be the same as `scrollTop`? I will double check and optimize if possible) `topTooClose`: false (50 is not less than 10, so we do not need to add more children at the top) `topTooFar`: true (50 is greater than 10 times 2, so we current have too many children at the top (and accordingly, not enough at the bottom)) `adjust`: increment function `clampFirstRenderedChildIdx`: function that clamps between 0 and 11. 11 because the total child count is 20, and the `renderedChildCount` is the number of visible children plus 2 (notionally, one on each end) which is 9. See line 39 where `renderedChildCount` is computed. `newFirstRenderedChildIdx`: first we divide 50 by 10 to get 5. Then subtract 2 (which is what you're asking about). Then call the `adjust` function which increments to 3. Then that passes the `clamp` and 3 is the final result.
> 
> Remember though, this is an array index, so this 3 refers to the _fourth_ child. And remember in this example that the fifth child is the first one visible at the new scroll position. So this result in one child outside of view above the visible area, then the five visible children, and then one child outside of view below the visible area. The children outside of view on either end are so that as you scroll there's something there to see rather than just immediately exposing the spacer.
> 
> TL;DR: I think we subtract 1 because array indexes are zero-based, and 1 more because we want the first rendered child to be the one above the first one you can actually see.

Thanks for taking the time to put in such a detailed response, it's much appreciated. I was thinking about indexing, but my apparently tired brain wasn't counting the child outside of view. The rest of the explanation was also a huge help in solidifying what's going on.